exasol/ai/mcp/server/keyword_search.py

```diff
@@ -0,0 +1,189 @@
+import re
+from collections import Counter
+from collections.abc import (
+    Generator,
+    Iterable,
+)
+from functools import cache
+from itertools import chain
+from math import log
+from typing import Any
+
+import numpy as np
+import spacy
+
+
+@cache
+def _get_word_split_pattern() -> re.Pattern:
+    r"""
+    The pattern splits camel case and snake case words into multiple words. However,
+    the camel case splitting only works for ASCII characters. More precisely, it will
+    not work if a word starts or ends with a diacritic or non-latin character.
+    The issue can be fixed using the regex library and the pattern
+    (?<=\p{Ll})(?=\p{Lu})|_+
+    """
+    pattern = r"(?<=[a-z])(?=[A-Z])|_+"
+    return re.compile(pattern, flags=re.MULTILINE)
+
+
+@cache
+def _get_word_extract_pattern() -> re.Pattern:
+    pattern = r"\w+"
+    return re.compile(pattern, flags=re.MULTILINE | re.UNICODE | re.IGNORECASE)
+
+
+@cache
+def _get_nlp(model_name: str):
+    return spacy.load(
+        model_name, enable=["tok2vec", "tagger", "attribute_ruler", "lemmatizer"]
+    )
+
+
+def _extract_lemmatized_words(
+    sentences: Iterable[str], model_name: str
+) -> Generator[None, None, str]:
+    nlp = _get_nlp(model_name)
+    for sentence in sentences:
+        text = _get_word_split_pattern().sub(" ", sentence)
+        doc = nlp(text)
+        for token in doc:
+            if not token.is_punct and not token.is_stop:
+                yield token.lemma_.lower()
+
+
+def _extract_raw_words(sentences: Iterable[str]) -> Generator[None, None, str]:
+    for sentence in sentences:
+        for match in _get_word_extract_pattern().finditer(
+            _get_word_split_pattern().sub(" ", sentence)
+        ):
+            yield match.group().lower()
+
+
+def extract_words(sentences: Iterable[str], model_name: str | None = None) -> list[str]:
+    """
+    Extracts words from all texts in the provided collection.
+    All words in camel case or snake case get split into multiple words.
+    All extracted words returned lowercased.
+    If a spaCy model name is provided the words get lemmatized.
+    """
+    if model_name:
+        return list(_extract_lemmatized_words(sentences, model_name))
+    return list(_extract_raw_words(sentences))
+
+
+def get_match_scores(corpus: list[list[str]], keywords: list[str]) -> list[float]:
+    """
+    For each text in the corpus assigns a keyword matching score.
+    Computes the score as the cosine similarity between the TF-IDF vectors
+    of the keywords and each of the text in the corpus.
+
+    Here is an implementation of the vanilla TF-IDF algorithm. Alternatively, the
+    scikit-learn `TfidfVectorizer` could be used. This would allow using additional
```

Comment:
Actually, I would prefer to use scikit learn especially because, you also implemented k-means. Scikit learn is well maintained package and simplifies here the code a lot. Also the cosine distance available there, if you want to use tok2vec

Reply:
ok, sounds ok for me.